lib/modules.nix

```diff
@@ -105,6 +105,10 @@ rec {
                   # when resolving module structure (like in imports). For everything else,
                   # there's _module.args. If specialArgs.modulesPath is defined it will be
                   # used as the base path for disabledModules.
+                  #
+                  # `specialArgs.class`:
+                  # A nominal type for modules. When set and non-null, this adds a check to
+                  # make sure that only compatible modules are imported.
                   specialArgs ? {}
```

Comment:
While I generally like the module class concept, I think it should be thought about a bit more. Here's some questions:
1. What about submodules? Should they be able to use classes too? If yes, what should the classes be called?
2. If we have lots of classes, would we have conflicts? Would we need a convention to avoid them or resolve them? Especially in light of third-party code interacting with each other, not just within Nixpkgs
3. What do these classes exactly do? Can we formally write down the semantics? Currently they don't have any changed behavior other than throw an error in some cases so this might not be that hard.
4. Related: Would we want versioning for this, now or in the future? This is made for third-parties where the Nixpkgs version that gets used can vary greatly, would we need to attach a NixOS release number to the class? Would we need a way for third-parties to provide different modules depending on the NixOS release and the class itself?  
5. Could there be false positives, where users _want_ to have a module shared between two differing classes? I guess this is probably fine since modules can always omit class for the check not to happen

Maybe I'm overly-anxious here, just trying to think of the possibilities

Reply:
> 1. What about submodules? Should they be able to use classes too?

This should work the same for submodules.

> If yes, what should the classes be called?

There's basically no different. `class` is still `class` and the values are for the module author to pick. If you use NixOS as a submodule, it's already set for you. If you use a submodule as a record type with no behavior or reuse, you don't need a `class`. If you use a submodule as a place to import modules, you can pick your own name for it. This would be rare in NixOS, but flake-parts would have e.g. `"flake-parts"` and `"flake-parts-perSystem"`, but for example `haskell-flake` has its own class of modules to be imported into haskell project submodules.

> 2\. If we have lots of classes, would we have conflicts? Would we need a convention to avoid them or resolve them? Especially in light of third-party code interacting with each other, not just within Nixpkgs

I've considered this, and decided against acting on this, as it is quite unlikely to cause real issues. Let's keep it simple.




> 3\. What do these classes exactly do? Can we formally write down the semantics?

They restrict which modules can be imported. A module can be imported when its class is `null` or equal to the `class` specified by through `specialArgs`.



> 4\. Related: Would we want versioning for this, now or in the future?

Currently we don't. In the future, we could, if the need arises.
Modules already have a lot of options to adapt to the `specialArgs` and `options` environment they're in, including the possibility to error out. I don't think this functionality needs to be coupled to the `class` feature.



> 5\. Could there be false positives, where users _want_ to have a module shared between two differing classes? I guess this is probably fine since modules can always omit class for the check not to happen

That's also my reasoning. Omission of a class is a feature, not a bug.



> Maybe I'm overly-anxious

I hope I've convinced you not to be :)